### DISCUSSION TITLE
Problem downloading the en_core_web_trf model on AWS EC2 Ubuntu Instance

### DISCUSSION THREAD

**Discussion by ansh-lehri** [explosion/spaCy#8672] [Help: Installation] [answered]
I am trying to download the en_core_web_trf model in the EC2 Ubuntu instance but it is not getting downloading. The system is not throwing any error rather it is completely downloading the 460MB "en_core_web_trf" model but is not finally collecting required dependencies and creating wheels.

Commands used to install spaCy and en_core_web_trf were:

`pip install -U spacy`    --- to download spaCy
`python3 -m spacy download en_core_web_trf`   ---- to download en_core_web_trf model 

On executing the above command following process happens and gets killed or stopped after this.

_Collecting en-core-web-trf==3.1.0
  Downloading https://github.com/explosion/spacy-models/releases/download/en_cor                                                                                        e_web_trf-3.1.0/en_core_web_trf-3.1.0-py3-none-any.whl (460.2 MB)
     |████████████████████████████████| 460.2 MB 43.7 MB/s eta 0:00:01_

On doing  `python3 -m spacy info`, I get:

**spaCy version     3.1.0
Location             /home/ubuntu/environment/try/lib/python3.8/site-packages/spacy
Platform             Linux-5.4.0-1045-aws-x86_64-with-glibc2.29
Python version   3.8.10
Pipelines**

Even when no error is thrown while downloading the "trf" model, it is still not getting added into the spaCy pipelines.

I am not able to understand what is going wrong here. Is there a problem with the EC2 instance I am using or some problem in commands?

EC2 configurations:

Family          Type          vCPUs         Memory(GiB)         Instance Storage(GB)          Network Performance            IPv6 Support  
t2               t2.micro           1                      1                               EBS(only)                     Low to Moderate                      Yes

Please help to rectify this issue.

**Comment by polm** [comment, marked as answer]
Sorry you're having trouble with this, we haven't heard of a problem like this before.

Do other models work?

Does the model show up in `pip freeze`?

Is there a directory with the model name in `/home/ubuntu/environment/try/lib/python3.8/site-packages/`?

**Reply by ansh-lehri** [reply]
Hell @polm, I tried your points and it seems like the other models are working quite fine.

I installed the "en_core_web_sm" model and it is showing up in the spaCy pipelines as well.


**spaCy version    3.1.0
Location         /home/ubuntu/environment/try/lib/python3.8/site-packages/spacy
Platform         Linux-5.4.0-1045-aws-x86_64-with-glibc2.29
Python version   3.8.10
Pipelines        en_core_web_sm (3.1.0)**


Also, I did pip freeze but the "trf" model is not showing up but the "sm" model does.
This is the list of packages shown by "pip freeze":

blis==0.7.4
catalogue==2.0.4
certifi==2021.5.30
chardet==4.0.0
click==7.1.2
cymem==2.0.5
**en-core-web-sm==3.1.0**
idna==2.10
Jinja2==3.0.1
MarkupSafe==2.0.1
murmurhash==1.0.5
numpy==1.21.0
packaging==21.0
pathy==0.6.0
preshed==3.0.5
pydantic==1.8.2
pyparsing==2.4.7
requests==2.25.1
smart-open==5.1.0
spacy==3.1.0
spacy-legacy==3.0.8
srsly==2.4.1
thinc==8.0.7
tqdm==4.61.2
typer==0.3.2
typing-extensions==3.10.0.0
urllib3==1.26.6
wasabi==0.8.2

I also checked the directory you mentioned above but again no "trf" model is listed there as well.

Is there any sort of problem because of the EC2 instance type I am using? Because I trained my Custom NER Transformer model on Google Colab using GPUs and it worked fine. Then I wanted to test if the transformer model could work without GPU and I performed the same steps(installed spacy, installed en_core_web_trf, installed torch, torchvision, torchaudio all for CPUs) in new session of the Colab and tested the model and it worked fine.

Commands to install spacy and en_core_web_trf model were basically the same as I am doing in EC2.
So can there be any issue with the instance type I am using or some other issue?

**Reply by polm** [reply]
Without a more specific error it's hard to say what's going wrong. Maybe you don't have enough hard drive space? I'd have to check to be sure but I wouldn't be surprised if you need twice the space of the model at install time (maybe it copies the data while unzipping or something). 

When you say the process is killed or stopped, how do you know that? Is there some more detailed error?

Also note that it looks like t2 instances don't have a GPU, so using the Transformer models will probably be really slow and not a good idea.

**Reply by ansh-lehri** [reply]
I am just guessing that process is getting killed or stopped because no error is shown. As for GPU and runtime, slow processing is not a problem and I am using the transformer-based NER model(because of better results) for a college project and speed is not a problem. 

As for disk space, that might be the case

**Reply by ansh-lehri** [reply]
Hello @polm You were right, the problem was due to limited space on free-tier. 
But taking any other instance will cost me much money as I am doing a project and plan to make it live within my college. Can you suggest economical ways of deploying spaCy transformer trained custom NER model? This is my first time deploying something of this sort and I am not able to figure it out.

This will really be a great help if you could.

**Reply by polm** [reply]
If you have hardware limitations my advice is to just not use the Transformer. It has the best accuracy but the other models are pretty good too.

**Reply by johachi** [reply]
I had the same/similar problem with en_core_web_trf not installing. The solution was the same (increase storage). In my case, I had to increase the swap size in Docker Desktop.

**Reply by imhans33** [reply]
@polm I am also trying to deploy my custom NER entity model in aws EC2. So how much RAM and storage would be required for this. Right now i am using a blank model to create custom model. But later i will try with transformer models too. Will medium or large instance will be sufficient ?

**Reply by polm** [reply]
@imhans33 

1. Don't ping individual maintainers.
2. Your question is only tangentially related to this one, you should have opened a new question.

Required memory depends on your model training settings, docs processed, and other factors, so it's hard to say. Our main advice is that for training with Transformers we recommend 10GB of GPU RAM, though for inference you can get away with less. For more detailed numbers I would recommend you run a pipeline on a local machine and measure memory usage.

**Reply by imhans33** [reply]
Is there any other platform apart from aws, azure etc where we can deploy spacy models and test it with few developers from different locations.

**Reply by polm** [reply]
spaCy is just Python code, you can put it on anything. I guess you could use HuggingFace hub? Like here:

https://huggingface.co/spacy/en_core_web_lg
